Classes/Command/BaseCommand.php

```diff
@@ -339,7 +339,26 @@ private function addCollections(Document &$document, array $collections): void
                 $documentCollection = GeneralUtility::makeInstance(Collection::class);
                 $documentCollection->setIndexName($collection);
                 $documentCollection->setLabel($collection);
-                $documentCollection->setOaiName((!empty($this->extConf['general']['publishNewCollections']) ? Helper::getCleanString($collection) : ''));
+                $setSpec = '';
+                if (!empty($this->extConf['general']['publishNewCollections'])) {
+                    // setSpec only allows unreserved characters (rfc2396).
+                    // alnum | "-" | "_" | "." | "!" | "~" | "*" | "'" | "(" | ")"
+                    // Here we are a little bit more restrictive because
+                    // some more unreserved characters are not allowed.
+                    // Convert whitespaces to dash.
+                    $setSpec = $collection;
+                    $setSpec = preg_replace('/[\s]/', '-', $setSpec);
+                    // Remove multiple dashes.
+                    $setSpec = preg_replace('/[-]{2,}/', '-', $setSpec);
+                    // Remove undesired characters.
+                    $setSpec = preg_replace('/[^\w:-]/', '', $setSpec);
+                    // A hierarchical setSpec consists of two or more
+                    // normal setSpec which are separated by colons.
+                    // Remove colons which don't separate hierarchical setSpec entries.
+                    $setSpec = trim($setSpec, ':');
+                    $setSpec = preg_replace('/:{2,}/', ':', $setSpec);
```

Comment:
Maybe don't allow colons for now until we support hierarchical sets?

Reply:
Support for hierarchical sets means that an OAI-PMH request with parameter `set=SETSPEC` must not only return entries from set "SETSPEC", but also entries from all sets "SETSPEC:*", right? That should not be very difficult to implement. Did I miss more cases?

Reply:
Yes, that's my understanding of the specification as well.